### PR TITLE
[WIP] Fix for phantom data? DO NOT MERGE

### PR DESCRIPTION
**Pull request by agramfort**
ping @britta-wstnr

### CODE REVIEW

File: mne/beamformer/_lcmv.py [hunk 1]
```diff
@@ -253,6 +253,8 @@ def make_lcmv(info, forward, data_cov, reg=0.05, noise_cov=None, label=None,
                 tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
                 eig_vals, eig_vecs = linalg.eig(np.dot(linalg.pinv(tmp),
                                                        np.dot(Wk, Gk)))
+                eig_vals = np.real(eig_vals)
+                eig_vecs = np.real(eig_vecs)
```
Comment (agramfort) [inline mne/beamformer/_lcmv.py]: this is needed to avoid complex numbers in filters

File: mne/beamformer/_lcmv.py [hunk 2]
```diff
@@ -407,8 +409,9 @@ def _prepare_beamformer_input(info, forward, label, picks, pick_ori):
         raise ValueError('Normal orientation can only be picked when a '
                          'forward operator with a surface-based source space '
                          'is used.')
-    # Restrict forward solution to selected channels
-    info_ch_names = [ch['ch_name'] for ch in info['chs']]
+    # Restrict forward solution to selected channels (without bads)
+    info_ch_names = [ch['ch_name'] for ch in info['chs']
+                     if ch['ch_name'] not in info['bads']]
```
Comment (agramfort) [inline mne/beamformer/_lcmv.py]: don't we need to remove bads from the filters?

File: mne/beamformer/_lcmv.py (hunk 1 above)
Comment (britta-wstnr) [inline mne/beamformer/_lcmv.py]: yes, this is what I had in mind, too

Comment (britta-wstnr) [inline mne/beamformer/_lcmv.py]: or, rather actually - it should be enough and probably more save to do:
```
denom = np.real(np.sqrt(tmp))
```

Comment (agramfort) [inline mne/beamformer/_lcmv.py]: do we have a mathematical reference I can read? a paper discussing this issue?

Comment (agramfort) [inline mne/beamformer/_lcmv.py]: @wmvanvliet any thought on this?

Comment (britta-wstnr) [inline mne/beamformer/_lcmv.py]: After first investigations on the phantom data set: the problem only arises if mags and grads are combined. So maybe the problem lies there?

**Comment by agramfort** [comment]
combining sensors should not be a problem if whitening is done with a noise
cov

File: mne/beamformer/_lcmv.py (hunk 1 above)
Comment (britta-wstnr) [inline mne/beamformer/_lcmv.py]: As a reference for the orientation selection in unit-noise-gain beamformers, there is Sekihara & Nagarajan 2008.

Comment (wmvanvliet) [inline mne/beamformer/_lcmv.py]: Why do we have complex data in the first place?

Comment (wmvanvliet) [inline mne/beamformer/_lcmv.py]: (and what's the "Phantom" data/)

Comment (britta-wstnr) [inline mne/beamformer/_lcmv.py]: The complex eigenvalues seem to relate to a highly rank-deficient covariance matrix (rank 74 with 306 channels) and whitening of the leadfield (computing the whitener on the noise covariance matrix)

**Comment by agramfort** [comment]
*@wmvanvliet I am sharing with you a dropbox folder.*

**Comment by wmvanvliet** [comment]
neat

File: mne/beamformer/_lcmv.py (hunk 1 above)
Comment (wmvanvliet) [inline mne/beamformer/_lcmv.py]: Two vectors in `tmp` here all lie on the same plane. This is probably why we end up with complex numbers.

**Comment by agramfort** [comment]
let me know if you understand better what's going on

**Comment by wmvanvliet** [comment]
I think @britta-wstnr is correct and that it is related to rank deficient covariance matrices. Does the problem go away with more regularization?

**Comment by agramfort** [comment]
@britta-wstnr <https://github.com/britta-wstnr> says it doesn't. What is
not clear to me is why mathematically eig shall return a real output if the
input is not guaranteed to be symmetric. Using eigh does not seem to be the
solution.

**Comment by wmvanvliet** [comment]
I'll have another look tomorrow.

**Comment by britta-wstnr** [comment]
Seems one can also provoke this behavior by choosing a bad noise covariance matrix - without the actual data covariance matrix being rank-deficient - I'll look deeper into that this afternoon.

**Comment by britta-wstnr** [comment]
Update on some ideas I followed up with today:
1) the ``somato`` data set does not show the same problems, even though the covariance matrix is rank deficient (rank 64 with 306 channels). 
2) voxels with complex eigenvalues are not clustered in a systematic way (i.e., not center of head etc.)
3) setting (stupidly) ``noise_cov = data_cov`` provokes complex eigenvalues 

Not sure, if we can draw any conclusions from that? Maybe it could point towards whitening going wrong?

**Comment by agramfort** [comment]
just a remark somato uses a bem. With phantom it's a sphere model

**Comment by britta-wstnr** [comment]
Indeed, somato crashes with a sphere model, too.

**Comment by britta-wstnr** [comment]
Checked the leadfield rank (per voxel): the sphere model leadfield has only rank 2 for some voxels (as expected, I guess), but these are not necessarily similar with the voxels that lead to complex eigenvalues.

**Comment by agramfort** [comment]
as you can use MNE leadfield in fieldtrip can you test if it works there?
and if so understand what the pb is? thx

**Comment by britta-wstnr** [comment]
Yes, sure!

**Comment by britta-wstnr** [comment]
First insights from parallel Python/MATLAB debugging: 
1. Fieldtrip's unit-noise gain beamformer returns complex values as well with the phantom data
2. Since Fieldtrip does no whitening, the results differ
3. If I bypass whitening in MNE, results still differ. This can be hunted down to MATLAB's pseudo-inverse computation returning different results on full-rank matrices than the normal inverse would do - ``scipy.linalg.inv`` and ``scipy.linalg.pinv`` return the same inverse in this case.

**Comment by agramfort** [comment]
pinv has a tol param. I would look into this.

**Comment by agramfort** [comment]
this was not the right fix see #4568

**Pull request closed without merging**